ndnsd/discovery/service-discovery.cpp

```diff
@@ -0,0 +1,374 @@
+                  m_fileProcessor.getAppPrefix(), 1600_ms,
+                  std::bind(&ServiceDiscovery::processSyncUpdate, this, _1))
+  , m_discoveryCallback(discoveryCallback)
+{
+  setUpdateProducerState();
+  setInterestFilter(m_producerState.applicationPrefix);
+
+  // listen on reload prefix as well.
+  setInterestFilter(ndnsd::discovery::NDNSD_RELOAD_PREFIX);
+}
+
+void
+ServiceDiscovery::setUpdateProducerState(bool update)
+{
+  NDNSD_LOG_INFO("Setting/Updating producers state: ");
+  if (update)
+  {
+    m_fileProcessor.processFile();
+  }
+  m_producerState.serviceName = m_fileProcessor.getServiceName();
+  m_producerState.applicationPrefix = m_fileProcessor.getAppPrefix();
+  m_producerState.serviceLifetime = m_fileProcessor.getServiceLifetime();
+  m_producerState.publishTimestamp = ndn::time::system_clock::now();
+  m_producerState.serviceMetaInfo = m_fileProcessor.getServiceMeta();
+}
+
+ndn::Name
+ServiceDiscovery::makeSyncPrefix(ndn::Name& service)
+{
+  ndn::Name sync("/discovery");
+  sync.append(service);
+  return sync;
+}
+
+uint8_t
+ServiceDiscovery::processFalgs(const std::map<char, uint8_t>& flags, const char type)
+{
+  auto key = flags.find(type);
+  if (key != flags.end())
+  {
+    return flags.find(type)->second;
+  }
+  else
+  {
+    NDN_THROW(Error("Flag type not found!"));
+    NDNSD_LOG_ERROR("Flag type not found!");
+  }
+}
+
+void
+ServiceDiscovery::producerHandler()
+{
+  auto& prefix = m_producerState.applicationPrefix;
+  NDNSD_LOG_INFO("Advertising service under Name: " << prefix);
+  doUpdate(prefix);
+  run();
+}
+
+void
+ServiceDiscovery::consumerHandler()
+{
+  NDNSD_LOG_INFO("Requesting service: " << m_serviceName);
+  run();
+}
+
+void
+ServiceDiscovery::run()
+{
+  m_face.processEvents();
+}
+
+void
+ServiceDiscovery::stop()
+{
+  m_face.shutdown();
+  m_face.getIoService().stop();
+}
+
+void
+ServiceDiscovery::setInterestFilter(const ndn::Name& name, const bool loopback)
+{
+  NDNSD_LOG_INFO("Setting interest filter on: " << name);
+  m_face.setInterestFilter(ndn::InterestFilter(name).allowLoopback(false),
+                           std::bind(&ServiceDiscovery::processInterest, this, _1, _2),
+                           std::bind(&ServiceDiscovery::onRegistrationSuccess, this, _1),
+                           std::bind(&ServiceDiscovery::registrationFailed, this, _1));
+}
+
+void
+ServiceDiscovery::processInterest(const ndn::Name& name, const ndn::Interest& interest)
+{
+
+  NDNSD_LOG_INFO("Interest received: " << interest.getName());
+  auto interestName = interest.getName();
+
+  // check if the interest is for service detail or to update the service
+  if (interestName == NDNSD_RELOAD_PREFIX)
+  {
+    NDNSD_LOG_INFO("Receive request to reload service");
+    // reload file.
+    m_fileProcessor.processFile();
+    setUpdateProducerState(true);
+    // send back the response
+    static const std::string content("Update Successful");
```

Comment:
Maybe define somewhere else if this is the standard response according to protocol